Sources/TuistEnvKit/Commands/CommandRegistry.swift

```diff
@@ -46,6 +53,11 @@ public final class CommandRegistry {
             if processArguments().dropFirst().first == "--help-env" {
                 parser.printUsage(on: stdoutStream)
             } else if let parsedArguments = try parse() {
+                let verbose = parsedArguments.get(verboseArgument) ?? false
+
+                System.shared.verbose = verbose
```

Comment:
I wonder how other CLIs do to pass the verbose argument down to the components that need it. The implementation is consistent with the idea of treating those instances as global instances, but as we start adding state, this might cause race conditions, especially when using threads. Not 100% convinced, but I don't know any other way to do it. Any other suggestions @tuist/core, @lakpa, @vytis, @andreacipriani ?

Reply:
The longer-term solution would be as I mentioned in the original comment, to use a fully-fledged logging system with support for logging levels. e.g. https://github.com/apple/swift-log. 

I implemented this as a intermediate solution so that people who are having errors can produce more detailed logs for us to investigate the issues.

Reply:
> The longer-term solution would be as I mentioned in the original comment, to use a fully-fledged logging system with support for logging levels. e.g.


What about going down that path instead? We can keep the `Printer` as our printer utility, but extend all its methods to accept an extra argument, `level`. That's not a breaking change, so we won't have to touch the calls to the printer.

Reply:
It's a much more involved change. Currently the `Printing` protocol which `Printer` conforms too has different functions for what are a mix of levels and different logging categories:

```swift
public protocol Printing: AnyObject {
    func print(_ text: PrintableString)
    func print(section: PrintableString)
    func print(subsection: PrintableString)
    func print(warning: PrintableString)
    func print(error: Error)
    func print(success: PrintableString)
    func print(errorMessage: PrintableString)
    func print(deprecation: PrintableString)
}
```

e.g. Error & Warnings are different logging levels, where as Section & Subsection are different categories. This won't work by simply adding a `level` option, it would just create more confusion.

What we need to do to support this would be to consolidate this API and move fully to use a system-wide logging util. I would eventually like us to do that, but we need to consider if we want an intermediate solution to enable better reporting now - if we don't then we can just wait and plan to do a more involved change when we fully understand all the pieces.

Reply:
> e.g. Error & Warnings are different logging levels, where as Section & Subsection are different categories. This won't work by simply adding a level option, it would just create more confusion.

Right, it'll require some API adjustments, which I didn't think about.

> What we need to do to support this would be to consolidate this API and move fully to use a system-wide logging util. I would eventually like us to do that, but we need to consider if we want an intermediate solution to enable better reporting now - if we don't then we can just wait and plan to do a more involved change when we fully understand all the pieces.

Could you expand on what you mean by a system-wide log util? Are we talking about the same thing when we talk about printing outputs and logging? I think the confusion is coming from there. I see everything as output to be passed to the printer, and you are perhaps differentiating between printing and logging.

Reply:
We would need a system which is capable of recieving all messages and then is able to forward those messages onto the printer if the verbosity of the system is high enough for it.

```
    trace
    debug
    info
    notice
    warning
    error
    critical
```

By default logging levels of `info` and above (upto `critical`) would be printed to the console.
When `--verbose` is enabled, it would also log `debug`.
When running through Xcode or with a debugger attached we can allow `trace` logging for even more fine-grain verbosity.

Typically we would just print the logs to the console, but we can also log to a file, or log with `os_log` to enable other tools such as `Console.app`.

Reply:
Got you! Thanks for the clarification. Up to you then, I'm more inclined towards implementing the approach that you are describing, but if you feel taking that intermediate step is a good idea go for it.

Reply:
I have closed the PR in favour of spending time on a more involved change to how we handle logging